src/v/resource_mgmt/tests/memory_sampling_tests.cc

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "resource_mgmt/memory_sampling.h"
+#include "test_utils/async.h"
+
+#include <seastar/core/memory.hh>
+#include <seastar/testing/thread_test_case.hh>
+#include <seastar/util/log.hh>
+
+#include <boost/test/tools/old/interface.hpp>
+#include <boost/test/unit_test.hpp>
+
+#include <chrono>
+#include <sstream>
+
+#ifndef SEASTAR_DEFAULT_ALLOCATOR
+
+SEASTAR_THREAD_TEST_CASE(test_no_allocs_in_oom_callback) {
+    auto oom_callback = memory_sampling::get_oom_diagnostics_callback();
+    size_t bytes_written = 0;
+    auto writer = [&bytes_written](std::string_view bytes) {
+        bytes_written += bytes.size();
+    };
+
+    auto before = seastar::memory::stats();
+    oom_callback(writer);
+    auto after = seastar::memory::stats();
+
+    // to make sure we are actually testing something
+    BOOST_REQUIRE_GT(before.mallocs(), 0);
+    BOOST_REQUIRE_GT(bytes_written, 0);
+    BOOST_REQUIRE_EQUAL(before.large_allocations(), after.large_allocations());
+    BOOST_REQUIRE_EQUAL(before.allocated_memory(), after.allocated_memory());
+    BOOST_REQUIRE_EQUAL(before.mallocs(), after.mallocs());
+}
+
+SEASTAR_THREAD_TEST_CASE(test_low_watermark_logging) {
+    seastar::logger dummy_logger("dummy");
+    std::stringstream output_buf;
+    dummy_logger.set_ostream(output_buf);
```

Comment:
This is cool, didn't know you could do this.

Now I see why the logger is configurable.